bgp-cp: use CiliumNode object and support ClusterPool v1/v2 IPAM

ldelossa opened the pull request:
This PR refactors bits of the BGP Control Plane to understand Cilium's Cluster Pool IPAM.

When Cluster Pool IPAM is configured the BGP control plane will configure itself to parse the CiliumNode object's PodCIDR blocks instead of the Kubernetes native Node resource's. 

When Cluster Pool IPAM is not configured, the normal Kubernetes Node resource is parsed instead.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

```release-note
Move the BGP Control Plane to utilize CiliumNode objects. This enable support for IPAM driven PodCIDR announcements. 
```

christarazi commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -10,17 +10,15 @@ import (
+	v2api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
 	"github.com/cilium/cilium/pkg/k8s/client/informers/externalversions"
+	v2 "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```
Nit: 
```suggestion
	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
	"github.com/cilium/cilium/pkg/k8s/client/informers/externalversions"
	listersv2 "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```

christarazi commented on pkg/bgpv1/mock/mocks.go:
```diff
@@ -6,28 +6,28 @@ package mock
 	"context"
 
-	v1 "k8s.io/api/core/v1"
+	v2listers "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```
File is not `goimports`'d

christarazi commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -150,19 +145,27 @@ func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...Controll
-		lo.FieldSelector = "metadata.name=" + nodetypes.GetName()
-	})
-	k8sfactory := informers.NewSharedInformerFactoryWithOptions(k8s.Client(), 0, selfTweakList)
+	// selfTweakList := informers.WithTweakListOptions(func(lo *metav1.ListOptions) {
```
Remove commented out code?

christarazi reviewed (changes requested):
Few minor comments, also reiterating Sebastian's concern as well.

ldelossa committed:
bgpv1-cp: introduce nodespecer interface

the BGP control plane needs to access the hosting Node's PodCIDR ranges.
the way these ranges are accessed can change dependent on the IPAM mode
Cilium is configured with.

in order to provide a singular API to the control plane, introduce an
interface which abstracts the differences between obtaining NodeSpec
information from a Kubernetes Node and a CiliumNode resource.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgp-cp: support multiple IPAM modes

this commit modifies the BGP control plane to take multiple IPAM modes
into account.

the BGP control plane will know determine which IPAM mode is being used
and obtain the PodCIDR range appropriately given the mode. Thus, this
commit patches in support for advertising dynamic cluster pools when
this IPAM mode is configured.

As of today, the supported IPAM modes are ClusterPool V1, ClusterPool V2
and Kubernetes.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
daemon: enforce IPAM types when initializing BGP control plane

this commit changes the agent's initialization to fatal if BGP control
plane is enabled along with an incompatible IPAM mode.

the bgp control plane docs are also updated to communicate this.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

gandro commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -143,49 +177,73 @@ type ControllerOpt func(*Controller)
+	case ipamOption.IPAMKubernetes:
+		var err error
+		selfTweakList := informers.WithTweakListOptions(func(lo *metav1.ListOptions) {
+			lo.FieldSelector = "metadata.name=" + nodetypes.GetName()
```
Why do you only filter for the node name here, and not also in the ClusterPool case?

ldelossa replied:
Ah, good point. 

Because I use the other factory to **also**  list BGPPeeringPolicies, I don't think I can place the node filter on it, or else I'd limit the listing resources by that "metadata.name" field, which the BGPPeeringPolicies do not have. Hmm. is there a way to say "Filter this resource, by this field" and leave other resources unfiltered?

ldelossa replied:
@gandro 
Ok, so I opt'd to filter the CiliumNode factory by metadata.name as well. 
I dug a bit into the code and the in-memory caches are not created until a constructor for the informers are called. 

In a subsequent PR I actually want to get rid of both the CiliumNode and Node informers in favor of grabbing the stores from the watchers.go file and using those to list from. It's a larger change tho and I'd like to keep that to its own PR if possible.

gandro commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -0,0 +1,91 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node %v: %v", s.Name, err)
+	}
+	if n.Spec.IPAM.PodCIDRs != nil {
```
```suggestion
	if len(n.Spec.IPAM.PodCIDRs) > 0 {
```

ldelossa replied:
Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

ldelossa replied:
Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

ldelossa replied:
Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

gandro replied:
True. Even if `PodCIDRs` is `[]string{}` (and not nil), we still return the correct value. Nevermind my nit then. Marking as resolved.

gandro commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -0,0 +1,91 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node %v: %v", s.Name, err)
+	}
+	if n.Spec.PodCIDRs != nil {
```
```suggestion
	if len(n.Spec.PodCIDRs) > 0 {
```

gandro commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -143,49 +177,73 @@ type ControllerOpt func(*Controller)
+	switch option.Config.IPAM {
+	case ipamOption.IPAMClusterPool:
+		fallthrough
+	case ipamOption.IPAMClusterPoolV2:
```
tiny nit (feel free to ignore)
```suggestion
	case ipamOption.IPAMClusterPool, ipamOption.IPAMClusterPoolV2:
```

ldelossa replied:
oh nice! forgot about that syntax, I like this more. :+1:

gandro reviewed (commented):
While I'm not familiar with the new bgp-cp, the IPAM part looks good to me, only one question that popped up (the rest are nits).

gandro commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -198,7 +256,7 @@ func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...Controll
 // A cancel of the provided ctx will kill the control loop along with the running
 // informers.
-func (c *Controller) Run(ctx context.Context) {
+func (c *Controller) Run(ctx context.Context, stop chan struct{}) {
```
The new stop channel argument seems unused?

ldelossa replied:
Believe its passed @ https://github.com/cilium/cilium/pull/19872/files#diff-663c8aba9492de183ebe045c8bf425d6a6a645970cd2a201a7073e9b21ffe8adR246

Look okay?

gandro replied:
The link doesn't work for me, but yes, sorry, looking at the full file it is used here:
https://github.com/cilium/cilium/blob/6687a6f3887eb7f6cd660d2d00cb57c1ba40f253/pkg/bgpv1/agent/controller.go#L277

ldelossa replied:
Cool. Marking resolved

gandro commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -1,4 +1,3 @@
-// SPDX-License-Identifier: Apache-2.0
```
Why is this removed?

ldelossa replied:
good catch, unintentional.

gandro reviewed (commented):
From what I know about the logic, this looks good to me. One minor question
